Add SVD function specification (#114)
data-apis/array-api: spec/API_specification/linear_algebra_functions.md

TODO

(function-svd)=

> ### svd(x, /, *, compute_uv=True, full_matrices=True)

Reviewer: Are we purposely deviating from NumPy's signature, in which `full_matrices` goes before `compute_uv`? Logically the choice here makes sense as `compute_uv` could be ignored, but I'd like to make sure this is a delibrate choice.

Author: These are keyword-only, and not positional, arguments, so the order in which they appear should not matter. Here, they are listed in alphabetical order.

Reviewer: Ah yes, thanks @kgryte! My builtin parser automatically discarded / and * ...

Computes the singular value decomposition `A = USV` of a matrix (or stack of matrices) `x`.

#### Parameters

-   **x**: _&lt;array&gt;_

    -   input array having shape `(..., M, N)` and whose innermost two dimensions form matrices on which to perform singular value decomposition. Must have a data type of either `float32` or `float64`.

-   **compute_uv**: _bool_

    -   If `True`, compute the left and right singular vectors and return as `u` and `v`, respectively. Default: `True`.

-   **full_matrices**: _bool_

>     -   If `True`, compute full-sized `u` and `v`, such that `u` has shape `(..., M, M)` and `v` has shape `(..., N, N)`. If `False`, compute on the leading `K` singular vectors, such that `u` has shape `(..., M, K)` and `v` has shape `(..., N, K)` and where `K = min(M, N)`. This option must be ignored if `compute_uv` is `False`. Default: `True`.

Reviewer: In NumPy, `v.shape == (..., K, N)`, not `(..., N, K)`. I suppose we wanna follow NumPy here?

Author: Indeed. I updated below, but not here. PR should be fixed.

#### Returns

-   **out**: _Union\[ &lt;array&gt;, Tuple\[ &lt;array&gt;, ... ] ]_

>     -   if `compute_uv` is `False`, an array with shape `(..., K)` that contains the vector(s) of singular values of length `K`. For each vector, the singular values must be sorted in descending order by magnitude, such that `s[..., 0]` is the largest value, `s[..., 1]` is the second largest value, et cetera. The first `x.ndim-2` dimensions must have the same shape as those of the input `x`.

Reviewer: Perhaps to be consistent across the `linalg` spectrum (ref: #95), this should return a namedtuple `(s,)` too?

Reviewer 2: I think those are all mistakes. A tuple with a single array in it is kind of weird and unpythonic. Doesn't matter if it's a namedtuple or not.

The design consistency is quite hard to see across all these linalg PRs. We need the overview PR with design principles for all linalg functions first I think.

Reviewer: Sounds good. Yeah I left a bunch of questions yesterday about whether it should be a namedtuple or just an array when there is only one object to be returned.

Author: This should now be resolved due to splitting of `svd` into `svd` and `svdvals`.

>     -   if `compute_uv` is `False`, an array having shape `(..., K)` and containing the vector(s) containing the singular values. For each vector, the values must be sorted in descending order according to magnitude, such that `s[..., 0]` is the largest value, `s[..., 1]` is the second largest value, et cetera. The first `x.ndim-2` dimensions must have the same size as those of the input `x`.

Reviewer: Probably this is the style used throughout the spec, but I find the first sentence is a bit terse. How about this?
```suggestion
    -   if `compute_uv` is `False`, an array with shape `(..., K)` that contains the vector(s) of singular values of length `K`. For each vector, the singular values must be sorted in descending order by magnitude, such that `s[..., 0]` is the largest value, `s[..., 1]` is the second largest value, et cetera. The first `x.ndim-2` dimensions must have the same shape as those of the input `x`.
```

Author: @leofang Thanks for the suggestion!

    -   if `compute_uv` is `True`, a namedtuple `(u, s, v)` whose
    

>         -   first element must be an array whose shape depends on the value of `full_matrices` and contain unitary array(s) (i.e., the left singular vectors). The left singular vectors must be stored as columns. If `full_matrices` is `True`, the array must have shape `(..., M, M)`. If `full_matrices` is `False`, the array must have shape `(..., M, K)`, where `K = min(M, N)`. The first `x.ndim-2` dimensions must have the same size as those of the input `x`.

Reviewer: ```suggestion
        -   first element must be an array whose shape depends on the value of `full_matrices` and contain unitary array(s) (i.e., the left singular vectors). The left singular vectors must be stored as columns. If `full_matrices` is `True`, the array must have shape `(..., M, M)`. If `full_matrices` is `False`, the array must have shape `(..., M, K)`, where `K = min(M, N)`. The first `x.ndim-2` dimensions must have the same shape as those of the input `x`.
```

>         -   second element must be an array having shape `(..., K)` and contain the vector(s) containing the singular values. For each vector, the values must be sorted in descending order according to magnitude, such that `s[..., 0]` is the largest value, `s[..., 1]` is the second largest value, et cetera. The first `x.ndim-2` dimensions must have the same size as those of the input `x`.

Reviewer: ```suggestion
        -   second element must be an array with shape `(..., K)` that contains the vector(s) of singular values of length `K`. For each vector, the singular values must be sorted in descending order by magnitude, such that `s[..., 0]` is the largest value, `s[..., 1]` is the second largest value, et cetera. The first `x.ndim-2` dimensions must have the same shape as those of the input `x`.
```

>         -   third element must be an array whose shape depends on the value of `full_matrices` and contain unitary array(s) (i.e., the right singular vectors). The right singular vectors must be stored as rows (i.e., the array is the adjoint). If `full_matrices` is `True`, the array must have shape `(..., N, N)`. If `full_matrices` is `False`, the array must have shape `(..., K, N)` where `K = min(M, N)`. The first `x.ndim-2` dimensions must have the same size as those of the input `x`. 

Reviewer: ```suggestion
        -   third element must be an array whose shape depends on the value of `full_matrices` and contain unitary array(s) (i.e., the right singular vectors). The right singular vectors must be stored as rows (i.e., the array is the adjoint). If `full_matrices` is `True`, the array must have shape `(..., N, N)`. If `full_matrices` is `False`, the array must have shape `(..., K, N)` where `K = min(M, N)`. The first `x.ndim-2` dimensions must have the same shape as those of the input `x`. 
```

Reviewer: Note that the shape is correctly given as `(..., K, N)` here.

Author: Good catch!
